### PR TITLE
Hotfix/Transaction middleware

### PR DESCRIPTION
**Pull request by pattisdr**
# Purpose

https://openscience.atlassian.net/browse/OSF-4759
Issue interfering with transactional requests. If there is an exception thrown, the TokuTransactionMiddleware is skipping commands.rollback. If anything is written to the database before the exception, it is committed.

DRF creates a nice, properly formatted response if there is an exception.  Since the response is not None, the process_exception in the TokuTransactionMiddleware is never being called. 

# Changes

Checks to see if an exception is thrown (by looking at the status_code).  If it's greater than or equal to 400, the transaction is rolled back.  Otherwise, the transaction is committed.

### CODE REVIEW

File: tests/api_tests/base/test_middleware.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    
+    @mock.patch('api.base.middleware.commands')
+    def test_400_error_causes_rollback(self, mock_commands):
+
+        middleware = TokuTransactionsMiddleware()
+        mock_response = mock.Mock()
+        mock_response.status_code = 400
+        middleware.process_response(mock.Mock(), mock_response)
+
+        assert_is(mock_commands.rollback.assert_called_once_with(), None)
```
Comment (sloria) [inline tests/api_tests/base/test_middleware.py]: The `assert_is` is unnecessary here; we only to assert that rollback was called.

I'll make this change when I merge.

Comment (abought) [inline tests/api_tests/base/test_middleware.py]: You appear to be nesting two assert statements here, which could get confusing.

If you're just trying to determine whether a function was called, consider using `mock_commands.rollback.called` instead- it should return a boolean (then you can `assert_true(...)`)

File: tests/api_tests/base/test_middleware.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+from tests.base import ApiTestCase
+
+class TestMiddlewareRollback(ApiTestCase):
+    
+    @mock.patch('api.base.middleware.commands')
+    def test_400_error_causes_rollback(self, mock_commands):
+
+        middleware = TokuTransactionsMiddleware()
+        mock_response = mock.Mock()
+        mock_response.status_code = 400
```
Comment (abought) [inline tests/api_tests/base/test_middleware.py]: You have a test for failure- could you also had a second test for success?

File: tests/api_tests/base/test_middleware.py (hunk 1 above)
Comment (sloria) [inline tests/api_tests/base/test_middleware.py]: Actually, @pattisdr can you make this change along with the other test change Andy suggested?

**Comment by abought** [comment]
CR1a pass complete: sorry for the crosstalk, Steve. :)

Dawn: two requests for changes here- add one test and clean up another. No major changes, though I haven't personally run this on a particular test case (such as bulk API operations) yet.

**Comment by abought** [comment]
Looks good to me; CR1 clear. 

Since rollback and commit are mocked out, I'm honestly not sure *what* happens to the data at the end of this unit test- does it just go into limbo? But there's no test anomalies, and it wouldn't affect production, so *shrugs*.

**Pull request merged**